File: c/xrange_decoder.cc
#include "xrange_decoder.h"

#include "xrange_code.h"

namespace twim {

uint32_t XRangeDecoder::readNumber(XRangeDecoder* src, size_t max) {
  if (max == 1) return 0;
  size_t state = src->state;
  size_t offset = state & XRangeCode::kMask;
  size_t result = (offset * max + max - 1) / XRangeCode::kSpace;
  size_t low = result * XRangeCode::kSpace;
  size_t base = low / max;
  size_t freq = (low + XRangeCode::kSpace) / max - base;
  state = freq * (state / XRangeCode::kSpace) + offset - base;
  while (state < XRangeCode::kMin) state = (state << 1u) | src->readBit();
  src->state = state;
  return result;
}

XRangeDecoder::XRangeDecoder(std::vector<uint8_t>&& data)
    : data(std::move(data)),
      state(1u),
      pos(0) {
  for (size_t i = 0; i < XRangeCode::kBits; ++i) {
    state = (state << 1u) | readBit();
  }
}

size_t XRangeDecoder::readBit() {
  size_t offset = pos >> 3u;
  if (offset >= data.size()) return 0;
  return (data[offset] >> (pos++ & 7u)) & 1u;
}

}  // namespace twim

File: js/XRangeDecoder.js
import {b8, b11, b11m, b16} from "./Mini.js";

// See Java sources for magic values explanation.

/**
 * @return {number}
 */
let readBit = () => {
  let /** @type{number} */ byte = _data[_pos >> 3];
  return (byte >> (_pos++ & 7)) & 1;
};

/** @type{number} */
let  _state;
/** @type{number} */
let _pos;
/** @type{!Uint8Array} */
let _data;

/**
 * @param{!Uint8Array} data
 * @return {void}
 */
export let init = (data) => {
  _data = data;
  //_state = 1;
  //_pos = 0;
  //for (let /** @type{number} */ i = 0; i < 16; ++i) _state = (_state * 2) + readBit();
  _state = ((_data[0] + _data[1] * b8) | 0) + b16;
  _pos = 16;
};

/**
 * @param{number} max
 * @return{number}
 */
export let readNumber = (max) => {
  if (max < 2) return 0;
  let /** @type{number} */ offset = _state & b11m;
  let /** @type{number} */ result = (offset * max + max - 1) >> 11;
  let /** @type{number} */ low = result * b11;
  let /** @type{number} */ base = (low / max) | 0;
  let /** @type{number} */ freq = (((low + b11) / max) | 0) - base;
  _state = freq * (_state >> 11) + offset - base;
  while (_state < b16) _state = (_state * 2) | readBit();
  return result;
};
